File: LeetCode/1578/golang/lc1578.go
package lc1578

// Runtime complexity: O(n)
// Auxiliary space: O(1)
// Subjective level: easy
func minCost(colors string, neededTime []int) int {
	previousTime := neededTime[0]
	ans := 0
	for i := 1; i < len(colors); i++ {
		if colors[i-1] == colors[i] {
			// The time for cutting the faster of the two is added to the immediate answer.
			// The time for cutting the slower of the two remains to be considered for the
			// next cutting, if the next balloon is also the same color.

			// don't want to implement max/min for ints. Go has them only for float64 for some reason.
			if previousTime < neededTime[i] {
				ans += previousTime
				previousTime = neededTime[i]
			} else {
				ans += neededTime[i]
				// previousTime remains as it was.
			}
		} else {
			// Don't cut anything, but remember this balloon's time in case the next is also same color.
			previousTime = neededTime[i]
		}
	}
	return ans
}

File: LeetCode/1578/cpp/lc1578.cpp
#include <cassert>
#include <string>
#include <vector>

class Solution {
public:
  int minCost(std::string colors, std::vector<int>& neededTime) {
    int ans = 0;
    int time = neededTime[0];
    for (int i = 1; i < colors.size(); ++i) {
      if (colors[i-1] == colors[i]) {
        // One of the two balloons has to go.
        ans += std::min(time, neededTime[i]);
        time = std::max(time, neededTime[i]);
      }
      else {
        // Hold the current balloon in hand, in case next one is same color:
        time = neededTime[i];
      }
    }
    return ans;
  }
};

int main() {
  Solution s;
  {
    std::string colors = "abaac";
    std::vector<int> neededTime = {1, 2, 3, 4, 5};
    assert(s.minCost(colors, neededTime) == 3);
  }
  {
    std::string colors = "abc";
    std::vector<int> neededTime = {1, 2, 3};
    assert(s.minCost(colors, neededTime) == 0);
  }
  {
    std::string colors = "aabaa";
    std::vector<int> neededTime = {1, 2, 3, 4, 1};
    assert(s.minCost(colors, neededTime) == 2);
  }
}
